File: src/main/python/AoC2019_03.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 3
#

from typing import Iterator
from aoc import my_aocd

Coord = tuple[int, int]
Wire = list[Coord]


def _parse(inputs: tuple[str]) -> (Iterator[str], Iterator[str]):
    return ((_ for _ in inputs[i].split(",")) for i in range(2))


def _to_leg(start: Coord, instruction: str) -> Iterator[Coord]:
    direction = instruction[0]
    amount = int(instruction[1:])
    if direction not in {"R", "U", "L", "D"}:
        raise ValueError("Invalid input")
    for i in range(1, amount + 1):
        if direction == "R":
            yield (start[0] + i, start[1])
        elif direction == "U":
            yield (start[0], start[1] + i)
        elif direction == "L":
            yield (start[0] - i, start[1])
        else:
            yield (start[0], start[1] - i)


def _to_wire(start: Coord, instructions: Iterator[str]) -> Wire:
    wire = Wire()
    for instruction in instructions:
        for coord in _to_leg(start, instruction):
            wire.append(coord)
            start = coord
    return wire


def _get_wires(wire_instructions) -> (Wire, Wire):
    return (_to_wire((0, 0), instructions)
            for instructions in wire_instructions)


def part_1(inputs: tuple[str]) -> int:
    wire1, wire2 = _get_wires(_parse(inputs))
    return min(abs(x) + abs(y) for x, y in set(wire1) & set(wire2))


def part_2(inputs: tuple[str]) -> int:
    wire1, wire2 = _get_wires(_parse(inputs))
    return min(wire1.index(c) + 1 + wire2.index(c) + 1
               for c in set(wire1) & set(wire2))


TEST1 = """\
R8,U5,L5,D3
U7,R6,D4,L4
""".splitlines()
TEST2 = """\
R75,D30,R83,U83,L12,D49,R71,U7,L72
U62,R66,U55,R34,D71,R55,D58,R83
""".splitlines()
TEST3 = """\
R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51
U98,R91,D20,R16,D67,R40,U7,R15,U6,R7
""".splitlines()


def main() -> None:
    my_aocd.print_header(2019, 3)

    assert part_1(TEST1) == 6
    assert part_1(TEST2) == 159
    assert part_1(TEST3) == 135
    assert part_2(TEST1) == 30
    assert part_2(TEST2) == 610
    assert part_2(TEST3) == 410

    inputs = my_aocd.get_input(2019, 3, 2)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()

File: src/main/java/AoC2019_03.java
import static java.util.stream.Collectors.toList;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

public class AoC2019_03 extends AoCBase {
    
    private static final Position ORIGIN = Position.of(0, 0);
    
    private final Wire wire1;
    private final Wire wire2;

    private AoC2019_03(final List<String> input, final boolean debug) {
        super(debug);
        assert input.size() == 2;
        this.wire1 = Wire.fromString(input.get(0));
        this.wire2 = Wire.fromString(input.get(1));
    }

    public static AoC2019_03 create(final List<String> input) {
        return new AoC2019_03(input, false);
    }

    public static AoC2019_03 createDebug(final List<String> input) {
        return new AoC2019_03(input, true);
    }
    
    @Override
    public Integer solvePart1() {
        final Set<Position> coords1 = new HashSet<>(wire1.getCoordinates());
        return wire2.getCoordinates().stream()
                .filter(coords1::contains)
                .mapToInt(c -> c.manhattanDistance(ORIGIN))
                .min().getAsInt();
    }

    @Override
    public Integer solvePart2() {
        final Set<Position> coords1 = new HashSet<>(wire1.getCoordinates());
        return wire2.getCoordinates().stream()
                .filter(coords1::contains)
                .mapToInt(c -> wire1.steps(c) + wire2.steps(c))
                .min().getAsInt();
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2019_03.createDebug(TEST1).solvePart1() == 6;
        assert AoC2019_03.createDebug(TEST2).solvePart1() == 159;
        assert AoC2019_03.createDebug(TEST3).solvePart1() == 135;
        assert AoC2019_03.createDebug(TEST1).solvePart2() == 30;
        assert AoC2019_03.createDebug(TEST2).solvePart2() == 610;
        assert AoC2019_03.createDebug(TEST3).solvePart2() == 410;

        final Puzzle puzzle = Aocd.puzzle(2019, 3);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2019_03.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2019_03.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST1 = splitLines(
            "R8,U5,L5,D3\r\n" +
            "U7,R6,D4,L4"
    );
    private static final List<String> TEST2 = splitLines(
            "R75,D30,R83,U83,L12,D49,R71,U7,L72\r\n" +
            "U62,R66,U55,R34,D71,R55,D58,R83"
    );
    private static final List<String> TEST3 = splitLines(
            "R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51\r\n" +
            "U98,R91,D20,R16,D67,R40,U7,R15,U6,R7"
    );
    
    @RequiredArgsConstructor
    private static final class Wire {
        @Getter
        private final List<Position> coordinates;
        
        public static Wire fromString(final String string) {
            final List<Position> wireCoordinates = new ArrayList<>();
            Position start = ORIGIN;
            for (final Instruction instruction : toInstructions(string)) {
                Position coord = null;
                for (int i = 1; i < instruction.amount + 1; i++) {
                    coord = start.translate(instruction.direction, i);
                    wireCoordinates.add(coord);
                }
                start = coord;
            }
            return new Wire(wireCoordinates);
        }
       
        public Integer steps(final Position coord) {
            return this.getCoordinates().indexOf(coord) + 1;
        }

        private static List<Instruction> toInstructions(final String input) {
            return Stream.of(input.split(","))
                    .map(Instruction::fromString)
                    .collect(toList());
        }
    }
    
    @RequiredArgsConstructor
    private static final class Instruction {
        private final Direction direction;
        private final int amount;
        
        public static Instruction fromString(final String str) {
            final Direction direction = Direction.fromString(str.substring(0, 1));
            final int amount = Integer.parseInt(str.substring(1));
            return new Instruction(direction, amount);
        }
    }
}